ntp-proto/src/clock.rs

```diff
@@ -8,5 +10,397 @@ pub trait NtpClock {
 
     fn now(&self) -> Result<NtpTimestamp, Self::Error>;
 
-    fn adjust_clock(offset: NtpDuration, freq_offset_ppm: f64) -> Result<(), Self::Error>;
+    fn set_freq(&self, freq: f64) -> Result<(), Self::Error>;
+    fn step_clock(&self, offset: NtpDuration) -> Result<(), Self::Error>;
+    fn update_clock(
+        &self,
+        offset: NtpDuration,
+        est_error: NtpDuration,
+        max_error: NtpDuration,
+        poll_interval: PollInterval,
+        leap_status: NtpLeapIndicator,
+    ) -> Result<(), Self::Error>;
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum ClockState {
+    StartupBlank,
+    // Needed when implementing frequency backups
+    #[allow(dead_code)]
+    StartupFreq,
+    MeasureFreq,
+    Spike,
+    Sync,
+}
+
+/// Controller responsible for actually
+/// deciding which adjustments to make based
+/// on results from the filtering and
+/// combining algorithms.
+#[derive(Debug, Copy, Clone)]
+pub struct ClockController<C: NtpClock> {
+    clock: C,
+    state: ClockState,
+    last_update_time: NtpInstant,
+    preferred_poll_interval: PollInterval,
+    poll_interval_counter: i32,
+    offset: NtpDuration,
+}
+
+pub enum ClockUpdateResult {
+    Ignore,
+    Step,
+    Slew,
+    Panic,
+}
+
+impl<C: NtpClock> ClockController<C> {
+    pub fn new(clock: C) -> Self {
+        clock.set_freq(0.).expect("Unable to set clock frequency");
+        Self {
+            clock,
+            state: ClockState::StartupBlank,
+            last_update_time: NtpInstant::now(),
```

Comment:
this default is not technically correct (no update was done now). But that doesn't matter for the logic?